// nba/lib/stats.js
"use strict";

var qs = require("querystring");
var debug = require("debug")("nba");
var template = require("nba-client-template");
var camelCase = require("camel-case");

var _require = require("./transforms");

var general = _require.general;
var players = _require.players;
var base = _require.base;
var lineups = _require.lineups;

var paramMap = {};
template.parameters.forEach(function (param) {
  paramMap[param.name] = param;
});

var transformMap = {
  playerProfile: general,
  playerInfo: general,
  playersInfo: players,
  teamStats: base,
  teamSplits: general,
  teamYears: base,
  playerSplits: general,
  shots: general,
  scoreboard: general,
  playByPlay: general,
  teamHistoricalLeaders: general,
  teamInfoCommon: general,
  commonTeamRoster: general,
  teamPlayerDashboard: general,
  lineups: lineups,
  playerTracking: general,
  homepageV2: general,
  assistTracker: general,
  playerStats: general,
  playerClutch: general,
  teamClutch: general,
  playerShooting: general,
  teamShooting: general
};

function makeStatsMethod(endpoint, transport) {

  var defaults = {};
  endpoint.parameters.forEach(function (param) {
    defaults[param] = paramMap[param]["default"];
  });

  var ccName = camelCase(endpoint.name);
  var transform = transformMap[ccName];
  // if (transform == null) {
  //   throw new Error(`No transform found for ${ccName}`);
  // }

  function statsMethod() {
    var query = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];
    var options = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

    var reqParams = Object.assign({}, defaults, query);

    debug("stats request", endpoint.url, reqParams);
    return transport(endpoint.url, reqParams).then(function (response) {
      if (response == null) return;

      // response is something like "GameID is required"
      if (typeof response === "string") throw new Error(response);

      return transform ? transform(response) : response;
    });
  }

  statsMethod.parameters = endpoint.parameters;
  statsMethod.defaults = defaults;
  return statsMethod;
}

function makeStatsClient(transport) {
  var client = {};
  template.stats_endpoints.forEach(function (endpoint) {
    client[camelCase(endpoint.name)] = makeStatsMethod(endpoint, transport);
  });
  client.withTransport = makeStatsClient;
  return client;
}

module.exports = makeStatsClient(require("./get-json"));
